src/mpi/ccl/cclcomm.c

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpiimpl.h"
+
+#define CUDA_ERR_CHECK(ret) if (unlikely((ret) != cudaSuccess)) goto fn_fail
+
+/*
+ * CCLcomm functions, currently tied to NCCL
+ */
+
+int MPIR_CCLcomm_init(MPIR_Comm * comm, int rank)
+{
+    int mpi_errno = MPI_SUCCESS;
+    cudaError_t ret;
+    int comm_size = comm->local_size;
+
+    MPIR_CCLcomm *cclcomm;
+    cclcomm = MPL_malloc(sizeof(MPIR_CCLcomm), MPL_MEM_OTHER);
+    MPIR_ERR_CHKANDJUMP(!cclcomm, mpi_errno, MPI_ERR_OTHER, "**nomem");
+
+    if(rank == 0) {
+        ncclGetUniqueId(&(cclcomm->id));
+    }
+
+    mpi_errno = MPIR_Bcast_impl(&(cclcomm->id), sizeof(cclcomm->id), MPI_INT, 0, comm,
+                                    MPIR_ERR_NONE);
+    MPIR_ERR_CHECK(mpi_errno);
+    
+    ret = cudaSetDevice(0); //TODO: Fix this for multi-GPU nodes
```

Comment:
probably want to move to coll invocation time and inherit whichever device the user-allocated buffer is located on. we could end up with multiple streams cached on the comm object.

Reply:
Fixed!